### PR TITLE
feat: unified homescreen with search and sort

### PR DESCRIPTION
**Pull request by cabreraalex**
# Description

Fix ZEN-306
Fix ZEN-302
Fix ZEN-284

### CODE REVIEW

**Review by Sparkier — CHANGES_REQUESTED** [review]
Overall really like these changes. Only minor things.

I know you wanted this for differentiation, but it looks like reports is selected:
image

Also here, I agree we need differentiation but I don't think this is it. Looks like it's selected to me:
image

The size difference here looks off IMHO:
image

Maybe just me but I liked the top banner better when it was lighter.

Is there something wrong for me here?
image

Personally don't like this as much but might just be taste. Generally mixing these changes in these big PRs makes it so that both take longer to get in so maybe better to separate out in the future:
image

Maybe we can move the search and filter stuff all to the top right generally? Might look better.

Do we want to remove this or did you just forget about it:
image

File: backend/zeno_backend/classes/homepage.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from zeno_backend.classes.project import Project, ProjectStats
+from zeno_backend.classes.report import Report, ReportStats
+
+
+class HomeTypeFilter(str, Enum):
+    """Type of entry to filter by."""
+
+    ALL = "ALL"
+    PROJECT = "PROJECT"
+    REPORT = "REPORT"
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/classes/homepage.py:15]: Maybe `EntryTypeFilter`? `TypeFilter` probably too generic.

File: backend/zeno_backend/database/insert.py [hunk 2]
```diff
@@ -393,18 +393,29 @@ async def system(
...
-        "INSERT INTO reports (name, owner_id, public) VALUES (%s,%s,%s);",
-        [name, user.id, False],
-    )
+    with Database() as db:
+        res = db.execute_return(
+            "SELECT id FROM reports WHERE name = %s AND owner_id = %s;",
+            [name, user.id],
+        )
+        if len(res) > 0:
+            return False
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/database/insert.py:412]: It does one step up

File: backend/zeno_backend/server.py [hunk 3]
```diff
@@ -505,41 +508,7 @@ def get_projects(current_user=Depends(auth.claim())):
...
-    def get_public_reports_details(reports_request: ReportsRequest, req: Request):
-        user = util.get_user_from_token(req)
-        reports = select.public_reports(reports_request)
-        report_stats = []
-        for rep in reports:
-            report_stats.append(select.report_stats(rep.id, user.id if user else None))
-        return ReportsDetails(
-            reports=reports, statistics=report_stats, num_reports=select.report_count()
-        )
+        return select.projects(user, HomeRequest())
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/server.py:511]: Yeah reports needs it

File: backend/zeno_backend/server.py [hunk 4]
```diff
@@ -550,7 +519,7 @@ def get_reports(current_user=Depends(auth.claim())):
         user = select.user(current_user["username"])
         if user is None:
             return Response(status_code=status.HTTP_401_UNAUTHORIZED)
-        return select.reports(user, ReportsRequest())
+        return select.reports(user, HomeRequest())
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/server.py:522]: Nope!

File: frontend/src/lib/components/report/Report.svelte [hunk 5]
```diff
@@ -92,8 +93,8 @@
 	</div>
 	<p class="mt-4 mr-2 text-sm w-full text-left overflow-y-auto flex-grow">
 		{#if report.description}
-			{report.description.slice(0, 160)}
-			{#if report.description.length > 160}
+			{report.description.slice(0, 100)}
+			{#if report.description.length > 100}
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply frontend/src/lib/components/report/Report.svelte:97]: Combined!

File: backend/zeno_backend/classes/homepage.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        limit: Limit of entries to return
+        search_string: String to search for in entries
+        type_filter: Type of entry to filter by
+        sort: Sort order for entries
+    """
+
+    user_name: str | None = None
+    offset: int = 0
+    limit: int | None = None
+    search_string: str = ""
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/classes/homepage.py:40]: I think for now it'll all be on home. can refactor/rename later if this changes?

File: backend/zeno_backend/classes/homepage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/classes/homepage.py:15]: Gonna leave as Home for now until we have a use for it elsewhere.

File: backend/zeno_backend/database/select.py [hunk 7]
```diff
@@ -128,14 +127,30 @@ def projects(user: User, projects_request: ProjectsRequest) -> list[Project]:
...
+            params += [
+                "%" + home_request.search_string + "%",
+                "%" + home_request.search_string + "%",
+            ]
+
+        if home_request.sort == HomeSort.POPULAR:
+            projects_query += sql.SQL(" ORDER BY total_likes DESC ")
+        elif home_request.sort == HomeSort.RECENT:
+            projects_query += sql.SQL(" ORDER BY updated_at DESC ")
+
```
[the remark this answers is not in the document]
Comment (cabreraalex) [reply backend/zeno_backend/database/select.py:148]: Eventually when we use limit and offset we don't want to get all entries in python - we'll get the top N * 2 for both projects and reports and then sort again the limited options in python was my thinking.

**Comment by cabreraalex** [comment]
It's ugly but I couldn't come up with anything better currently.

Screenshot 2023-10-31 at 16 35 41

IMO feel free to update design stuff however you want, if you don't find any big issues I'm down to merge and deploy to fix some of the homepage issues. We can then do a bigger refactor/clean up of the homepage.

**Comment by Sparkier** [comment]
Think it looks much better than before. Still feel like reports have more prominence with this highlighting which we should probably avoid but should get this in for now.

**Pull request merged**